4241 change agency type to enum

Benjamin-Couey opened the pull request:
Resolves #4241

### Description

This PR modifies the attribute `Partners::Profile.agency_type` to be a Rails enum that stores a code-string in the database. The `partners_profile.en.yml` localization file was also created so I18n may be used to convert from the code-string to a full, descriptive string. These translations were added to existing views.

The existing map of code-strings to descriptive strings, `Partners::AGENCY_TYPES`, was left as is because it is a dependency of the migration `20241122201255_cleanup_partner_agency_types.rb` and removing it would necessitate modifying this old migration, which would cause even worse problems. A warning comment was added to steer people away from using `AGENCY_TYPES` in the future and other parts of the code which referenced `AGENCY_TYPES` where updated to reference the new enum.

Lastly, a migration was created to convert existing values of `agency_type` to the specific values of the enum in cases where the existing value directly maps to one of the enum's values.

### Type of change

* New feature (non-breaking change which adds functionality)

### How Has This Been Tested?

Existing tests which referenced `agency_type` were updated to refer to the enum. No new tests were added.

I manually verified the translations worked in the pages for viewing and updating a partner's profile.

dorner commented:
Just a note about

> The existing map of code-strings to descriptive strings, Partners::AGENCY_TYPES, was left as is because it is a dependency of the migration 20241122201255_cleanup_partner_agency_types.rb

Honestly? We can just delete that migration from the codebase. It's already run on production and staging, and people's local databases shouldn't be affected.

dorner commented on app/models/partners/profile.rb:
```diff
@@ -95,6 +95,17 @@ class Profile < Base
     accepts_nested_attributes_for :served_areas, allow_destroy: true
 
     has_many_attached :documents
+    enum :agency_type, other: "OTHER", career: "CAREER", abuse: "ABUSE", bnb: "BNB",
```
Can this be reformatted so each value is on a single line?

dorner commented on app/models/partner.rb:
```diff
@@ -241,13 +246,14 @@ def agency_info
       zip_code: profile.zip_code,
       website: profile.website,
-      agency_type: (profile.agency_type == AGENCY_TYPES["OTHER"]) ? "#{AGENCY_TYPES["OTHER"]}: #{profile.other_agency_type}" : profile.agency_type
+      agency_type: (symbolic_agency_type == :other) ? "#{Partners::Profile.agency_types[:other]}: #{profile.other_agency_type}" : profile.agency_type
```
I'm assuming the values here are for display? Then I'd suggest

`agency_type: (symbolic_agency_type == :other) ? "Other: #{profile.other_agency_type}" : profile.agency_type`

dorner replied:
Actually - would we want to do `i18n.translate` for these values? As far as I can tell this is only used in the CSV export, so we should have readable strings here.

Benjamin-Couey replied:
I had originally left `profile.agency_type` because @cielf had a recollection that the type codes were used in the CSV and some of the full descriptive strings seemed quite long for a CSV column..

Below are some screenshots of the CSV under different conditions. Which looks best?

<details>
<summary>CSV with full descriptive strings</summary>

![PartnerCSVExportFullStrings](https://github.com/user-attachments/assets/d8c3c105-cf4b-460d-82a2-080a019b975e)

</details>

<details>
<summary>CSV with full descriptive strings (agency type column fully expanded)</summary>

![PartnerCSVExportFullStringsWideColumn](https://github.com/user-attachments/assets/9882f572-b145-47f9-b922-4bb77341a00e)

</details>

<details>
<summary>CSV with untranslated type codes</summary>

![PartnerCSVExportCodes](https://github.com/user-attachments/assets/c041a1e2-cb2e-4f5e-9ce8-f41bdb3d9599)

</details>

<details>
<summary>CSV with untranslated type codes, including `other_agency_type` for `:other`</summary>

![PartnerCSVExportCodesOtherType](https://github.com/user-attachments/assets/3ca93f02-b636-4845-8026-adc3d4b4d240)

</details>

dorner commented on app/views/partners/profiles/edit/_agency_information.html.erb:
```diff
@@ -8,7 +8,13 @@
+            <%= profile_form.input :agency_type,
+            # Symbolize keys, because simple_form will only do a I18n lookup for
+            # symbols
+              collection: Partners::Profile.agency_types.symbolize_keys.keys,
```
Probably a bit more performant to do `Partners::Profile.agency_types.keys.map(&:symbolize)`, but I may be nitpicking here.

Benjamin-Couey replied:
Made this change in the recent commits. For my own edification, why was my implementation less performant?

dorner replied:
Beacuse you go through the list of values twice, once to symbolize every key and then again to collect the keys from the entries. Like I said, it's a nitpick. :)

dorner commented on db/migrate/20250116180121_convert_partner_agency_types_to_enum.rb:
```diff
@@ -0,0 +1,77 @@
+    # Some agency types have trailing spaces -- need to get rid of them first
+    update 'UPDATE partner_profiles SET agency_type = TRIM(agency_type);'
+
+    profiles = Partners::Profile.all()
+    profiles.each do |profile|
```
```suggestion
    Partners::Profile.all.each do |profile|
```

Benjamin-Couey replied:
Made this change in the recent commits.

dorner commented on db/migrate/20250116180121_convert_partner_agency_types_to_enum.rb:
```diff
@@ -0,0 +1,77 @@
+      # Read the agency_type without casting so values not part of the enum aren't
+      # read as nil
+      current_agency_type = profile.read_attribute_before_type_cast(:agency_type)
+      puts profile.id
+      puts current_agency_type
```
Please remove these `puts` statements.

Benjamin-Couey replied:
Made this change in the recent commits.

dorner commented on db/migrate/20250116180121_convert_partner_agency_types_to_enum.rb:
```diff
@@ -0,0 +1,77 @@
+      current_agency_type = profile.read_attribute_before_type_cast(:agency_type)
+      puts profile.id
+      puts current_agency_type
+      if !current_agency_type.nil?
```
```suggestion
      if current_agency_type.present?
```

Benjamin-Couey replied:
Made this change in the recent commits.

dorner commented on db/migrate/20250116180121_convert_partner_agency_types_to_enum.rb:
```diff
@@ -0,0 +1,77 @@
+      if !current_agency_type.nil?
+        # If a profile has a descriptive string (ignoring case) as an agency type
+        # update it to the associated value code.
+        if( string_to_value_mapping.key?( current_agency_type.downcase ) )
```
```suggestion
        if string_to_value_mapping.key?(current_agency_type.downcase)
```

(similarly below - not sure why lint didn't pick this up)

Benjamin-Couey replied:
Hmm, for some reason the linter missed this file entirely. It only made corrections when I specified this file explicitly. Regardless, ran this through the inter in the recent commits.

dorner commented on spec/models/partner_spec.rb:
```diff
@@ -294,7 +294,7 @@
     let(:agency_zipcode) { "09980-7010" }
     let(:agency_website) { "bosco.example" }
-    let(:agency_type) { Partner::AGENCY_TYPES["OTHER"] }
+    let(:agency_type) { Partners::Profile.agency_types[:other] }
```
```suggestion
    let(:agency_type) { :other }
```

Benjamin-Couey replied:
Made this change, and also realized the test was improperly expecting the enum value (`OTHER`) which was fixed. Might need to update this again depending on what we choose for the CSV export.

dorner reviewed (changes requested):
Overall looks good! Some requests/comments.

cielf commented:
We'll need me to do some light manual testing when those are all addressed, methinks.

Benjamin-Couey commented:
> Just a note about
> 
> > The existing map of code-strings to descriptive strings, Partners::AGENCY_TYPES, was left as is because it is a dependency of the migration 20241122201255_cleanup_partner_agency_types.rb
> 
> Honestly? We can just delete that migration from the codebase. It's already run on production and staging, and people's local databases shouldn't be affected.

Alright, in the most recent commits I removed `20241122201255_cleanup_partner_agency_types.rb` and `Partners::AGENCY_TYPES`. I also copied the mappings from `20241122201255_cleanup_partner_agency_types.rb` to the migration added by this PR to preserve the relations and just in case someone hadn't run the now deleted migration.

Benjamin-Couey commented:
Alright, a bunch of checks are failing post merge. Working on resolving them.

Benjamin-Couey commented:
Alright, resolved the issues that were causing checks to fail. Now just waiting on @cielf and/or @dorner to make a decisions regarding [how we want to style the agency types when exported to csv](https://github.com/rubyforgood/human-essentials/pull/4931#discussion_r1923959192). After that, I can do another round of manual testing.

cielf commented:
Sorry I missed that-- this should be as seamless a change as possible for the users,  so keep putting out the full text, please.

cielf commented on config/locales/partners_profile.en.yml:
```diff
@@ -0,0 +1,52 @@
+agency_type_map: &agency_type_map
```
Functional issue:  "Other" is appearing at the top of the list of types when selecting.  It should be at the bottom.  Everything else should be in alphabetical order (by the full text) (though I note it isn't in staging --so the alpha thing could be a separate issue)

Benjamin-Couey replied:
@cielf alright this should be addressed in the most recent commit. Below is a screenshot demonstrating the new behavior.

<details>
<summary>Partner facing edit page with dropdown open</summary>

![PartnerFacingEditPageDropdownAlphabetized](https://github.com/user-attachments/assets/8228e140-b71b-41d1-8156-ced4b2ee79bd)

</details>

cielf reviewed (dismissed):
Looks pretty good functionally - except for the ordering of the selection list.  Please address that.

cielf commented:
Looks pretty good functionally on seed data.    We should probably run the migration against a local copy of prod data to make sure there aren't any hiccups, though.

dorner reviewed (approved):
All good on the technical side. @cielf is there any other testing you want to do before merging?

cielf commented:
Yup -- I'd like to run the migration against a copy of prod.   Probably tomorrow, but maybe not until Friday.
